### ISSUE TITLE
Problem with Pytest 2.8 and Django settings

### ISSUE THREAD

**Issue by kottenator** [pytest-dev/pytest#1059]
I have this problem: https://github.com/pytest-dev/pytest-django/issues/272

But it's not a bug of `pytest-django`, it caused by new Pytest 2.8 behaviour. I have `conftest.py`:

```py
from django.conf import settings

def pytest_configure():
    settings.configure(...)
```

Pytest 2.8 is trying to read properties of `settings` variable before Django is configured. This causes an error. I can do some workaround, but is it the correct way to solve it?

**Comment by blueyed** [comment]
I think it's a bug with pytest-django - see my latest comment: https://github.com/pytest-dev/pytest-django/issues/272#issuecomment-143040650.

Sorry for asking to report it against pytest in the beginning - it was not clear to me.

**Comment by nicoddemus** [comment]
@blueyed no problem!

This can be closed here then?

**Comment by kottenator** [comment]
I don't mind if you close this one. 

BTW, could you tell what's `pytest` trying to get from the `settings` variable?

**Comment by blueyed** [comment]
@kottenator 
It's not pytest doing something special with it, but accessing it in general will trigger the Django settings machinery / setup.
pytest-django's job is it to call e.g. `django.setup()` before that and/or prepare everything.
I am closing this issue here - I think the proper workaround for now would be to call `django.setup()`, instead of importing/using `settings` via `conf.settings`.  There are likely some issues/tickets in pytest-django.  Let's move the discussion over there.

**Comment by RonnyPfannschmidt** [comment]
I suspect fixture/hook discovery is involved

**Comment by blueyed** [comment]
JFI: turns out to be a problem with pytest indeed: https://github.com/pytest-dev/pytest/issues/1073

**Comment by kottenator** [comment]
@RonnyPfannschmidt - for example in Fabric, they were considering each Python function inside `fabfile.py` as a Fabric task. But then they have added `@task` decorator to distinguish utility functionas and Fabric tasks

**Comment by RonnyPfannschmidt** [comment]
@kottenator @hpk42 i suppose we could make pluggy consider inly decorated hook implementations if all of them are marked

@hpk any oppinion wrt intoducing a spec object in pluggy that could be used in each plugin to declare the actual plugin (then we could also use spec objects instead of classes to configure aspects of a plugin)

**Comment by nicoddemus** [comment]
Duplicate of #1073

**Comment by nicoddemus** [comment]
Created a PR at hpk42/pluggy#12. Once that is merged and a new `pluggy` is released, we must vendor it back into `pytest` before closing this.
